Allow polyfilling imperatively for a single style element

marchbox opened the pull request:
## Description

This change provides an alternative way for authors to apply polyfill: instead of letting the polyfill to lookup stylesheets and 
apply the polyfill, the second argument takes a `<style>` element and the function can apply polyfill based on its content.

This is particularly useful for components like custom elements, as often the stylesheets are added dynamically to the document, and the currently "auto" polyfilling makes it challenging to apply the changes at the right time.

## Related Issue(s)

https://github.com/oddbird/css-anchor-positioning/issues/228

jgerigmeyer commented on src/polyfill.ts:
```diff
@@ -443,13 +443,14 @@ async function position(rules: AnchorPositions, useAnimationFrame = false) {
+
   // fetch CSS from stylesheet and inline style
-  let styleData = await fetchCSS();
+  let styleData = el ? [{el, css: el.textContent ?? ''}] : await fetchCSS();
```
Elsewhere we're using `innerHTML`, but I'm not sure if there's a great argument for one over the other.

jgerigmeyer commented on src/polyfill.ts:
```diff
@@ -443,13 +443,14 @@ async function position(rules: AnchorPositions, useAnimationFrame = false) {
 }
 
-export async function polyfill(animationFrame?: boolean) {
+export async function polyfill(animationFrame?: boolean, el?: HTMLStyleElement) {
```
We should have done this from the beginning, but I think we should switch this from 2 argument to 1 options object. For backwards compatibility we can check if `typeof options === 'boolean'`, but otherwise going forward that will be both clearer and more flexible.

jgerigmeyer reviewed (commented):
I'm okay with this approach. I wonder if we want to accept an array of elements, and we could also support `<link>` tags as well as `<style>`? Basically using the same functionality as the existing polyfill, but allowing users to opt into a specified list of CSS sources.

@jamesnw @mmalerba What do you think?

jamesnw commented:
> I'm okay with this approach. I wonder if we want to accept an array of elements, and we could also support `<link>` tags as well as `<style>`? Basically using the same functionality as the existing polyfill, but allowing users to opt into a specified list of CSS sources.

I think an array that supports `<link>` tags as well as `<style>` makes a lot of sense. 

This does omit inline styles- could we add another boolean flag `includeInline` that defaults to true, and then adds the response of `fetch:fetchInlineStyles()`? Or since the inline style impact is scoped to just ones that include `anchor`, should we just always add the inline styles?

It would be nice to have a demo, but that's a bit of a challenge in the existing single page setup.

marchbox commented:
Thanks for the comments. I'll add them in.

> I think an array that supports <link> tags as well as <style> makes a lot of sense.
>
> This does omit inline styles- could we add another boolean flag includeInline that defaults to true, and then adds the response of fetch:fetchInlineStyles()? Or since the inline style impact is scoped to just ones that include anchor, should we just always add the inline styles?

Sounds good. I'll pass in the element array into both `fetchCSS()` and `fetchInlineStyle()` to build the `styleData` object.

marchbox commented:
I updated the code and it's ready for review:

* Changed the argument of `polyfill()` to accept an object for configuration, with backward compatibility
* You can pass in an array of elements for polyfilling, and imperative polyfill uses the same functions as auto polyfill
* Added demo and tests

Also, maybe I should rename "imperative" to "manual"? Let me know if "manual" sounds better and less confusing.

jamesnw reviewed (changes requested):
Excellent work, this is quite helpful! A couple quick notes-

- It looks like there are a few linting issues, you can run `npm run lint` to fix or surface.
- I do prefer `manual` over `imperative` slightly, but it's not a blocker.
- Is it a legitimate use case for people to pass in explicit `<style>` or `<link>` elements but want to polyfill all inline styles automatically? If so, that currently isn't possible. 
- If you click the imperative button, and then apply the polyfill to the whole page, the polyfill crashes. This is somewhat related to #255 as well as #91, so I'm not concerned here. 
- Do you have any concerns about running this on 2 discrete sets of elements separately? It seems like it should work, as long as there aren't references between the 2 sets of elements.

Overall, thanks for the contribution!

marchbox commented on src/fetch.ts:
```diff
@@ -52,12 +61,16 @@ function fetchInlineStyles() {
   const sources: Partial<StyleData>[] = [];
 
-  elements.forEach((el) => {
+  targetElements.forEach((el) => {
+    if (!(el instanceof HTMLElement)) {
+      return;
+    }
```
Ah definitely like the filter better! Thanks!

marchbox commented on index.html:
```diff
@@ -78,6 +84,77 @@
+          });
+        });
+      } else {
+        prepareImperativePolyfill();
```
Yeah sounds good! I'll update this.

marchbox commented:
> Is it a legitimate use case for people to pass in explicit <style> or <link> elements but want to polyfill all inline styles automatically? If so, that currently isn't possible.

I think that's a valid use case. I'll add this. My only concern is, if the `elements` already contains some elements that have inline styles, then you pass in `includeInlineStyles: false`, an author may expect the elements with inline styles in `elements` not get polyfilled. But this can probably be clarified by documentation. Or maybe we can give `includeInlineStyles` a different name for more clarity.

> Do you have any concerns about running this on 2 discrete sets of elements separately? It seems like it should work, as long as there aren't references between the 2 sets of elements.

Yeah this should work, and this is our intended use since multiple components in our library would need to manual polyfill separately. I also added some test coverage: https://github.com/oddbird/css-anchor-positioning/pull/256/commits/912c553c6e03502324ce9e4575e766321a4cff06.

The only caveat though, is that, if the elements in different sets rely on the same `anchor-name`, then all `polyfill()` call need to include the element that contains this `anchor-name` declaration. I'm not sure if there's a map somewhere that keeps a record of the anchor name and the generated custom property name. It would be nice to be able to reference an already transformed anchor name.

marchbox commented:
Added `includeInlineStyles`: https://github.com/oddbird/css-anchor-positioning/pull/256/commits/798561f9dbe8d38ac5c740d3fb3aaafd2de0b0b3

jamesnw commented:
> > Is it a legitimate use case for people to pass in explicit <style> or  elements but want to polyfill all inline styles automatically? If so, that currently isn't possible.
> 
> I think that's a valid use case. I'll add this. My only concern is, if the `elements` already contains some elements that have inline styles, then you pass in `includeInlineStyles: false`, an author may expect the elements with inline styles in `elements` not get polyfilled. But this can probably be clarified by documentation. Or maybe we can give `includeInlineStyles` a different name for more clarity.

I have a slightly different thought on how this could work that may make it more clear (either way, documentation will be useful). 

We could call it `discoverInlineStyles` or `transformInlineStyles` or ..., have it default to true, and separate it from whether or not `elements` are passed in. In other words, we default to always finding all the inline styles, but a user can opt out because they are passing in the elements they know have inline styles OR because they know they don't have any inline styles with `*anchor`.

Would that be simpler for users?

jgerigmeyer commented on index.html:
```diff
@@ -1001,6 +1119,87 @@ <h2>
+    right: anchor(--my-anchor-manual left);
+  }
+&lt;/style&gt;
+&lt;link rel="stylesheet" href="/anchor-manual.css" id="my-style-manual-link-el"&gt;
```
```suggestion
&lt;link rel="stylesheet" href="/anchor-manual.css" id="my-style-manual-link-el" /&gt;
```

jgerigmeyer commented on index.html:
```diff
@@ -1001,6 +1119,87 @@ <h2>
+
+&lt;script&gt;
+  polyfill({
+    useAnimationFrame: true,
```
Same question as above...

```suggestion
```

jgerigmeyer commented on index.html:
```diff
@@ -78,6 +85,117 @@
+
+        if (!SUPPORTS_ANCHOR_POSITIONING) {
+          polyfill({
+            useAnimationFrame: true,
```
Is there a reason to set `useAnimationFrame` here? I'd generally recommend people not use it unless they need to.

```suggestion
```

marchbox replied:
No particular reasons, I'll remove them.

jgerigmeyer commented on index.html:
```diff
@@ -1001,6 +1119,87 @@ <h2>
+    <section id="anchor-manual" class="demo-item" style="position: relative">
+      <h2>
+        <a href="#manual" aria-hidden="true">🔗</a>
+        Manually apply polyfill
```
I wonder if there's a clearer description for the intent of this change. Maybe:

```suggestion
        Polyfill applied to only specified styles
```

Or something along those lines?

marchbox replied:
How about "Manually apply polyfill to specific styles"?

jgerigmeyer commented on src/fetch.ts:
```diff
@@ -31,45 +31,64 @@ async function fetchLinkedStylesheets(
-    document.querySelectorAll('[style*="anchor"]');
+function fetchInlineStyles(elements?: HTMLElement[]) {
+  const elementsWithInlineAnchorStyles: HTMLElement[] = elements
+    ? elements.filter((el) =>
+        el?.matches?.(ELEMENTS_WITH_INLINE_ANCHOR_STYLES_QUERY),
+      )
```
It doesn't hurt anything, but I'm curious why the optional chaining is needed here.

```suggestion
    ? elements.filter((el) =>
        el.matches(ELEMENTS_WITH_INLINE_ANCHOR_STYLES_QUERY),
      )
```

marchbox replied:
I think I probably used them before I added type checking to make sure they will indeed be `HTMLElement`s. I'll remove these.

jgerigmeyer commented on src/fetch.ts:
```diff
@@ -31,45 +31,64 @@ async function fetchLinkedStylesheets(
-    inlineStyles.push({ el, css });
-  });
+  elementsWithInlineAnchorStyles
+    .filter((el) => el instanceof HTMLElement)
```
This is also more defensive than strictly necessary, right? I'm fine leaving it, just want to be sure I'm understanding the intent.

marchbox replied:
Since the previous check in `normalizePolyfillOptions()` in `polyfill.ts` only makes sure that `elements` is either an array or `undefined`, so I want to make sure no JS error would be thrown if an author acceidentally passed in a non `HTMLElement` array member.

jgerigmeyer commented on src/polyfill.ts:
```diff
@@ -443,13 +443,50 @@ async function position(rules: AnchorPositions, useAnimationFrame = false) {
+  // positions
+  useAnimationFrame?: boolean;
+
+  // An array of explictily targeted elements to polyfill
```
```suggestion
  // An array of explicitly targeted elements to polyfill
```

marchbox replied:
Nice catch!

jgerigmeyer commented on src/polyfill.ts:
```diff
@@ -443,13 +443,50 @@ async function position(rules: AnchorPositions, useAnimationFrame = false) {
+      : useAnimationFrameOrOption;
   const useAnimationFrame =
-    animationFrame === undefined
+    options.useAnimationFrame === undefined
       ? Boolean(window.UPDATE_ANCHOR_ON_ANIMATION_FRAME)
-      : animationFrame;
+      : options.useAnimationFrame;
```
In #245 there was a suggestion to convert this `window.UPDATE_ANCHOR_ON_ANIMATION_FRAME` option into a broader `window.ANCHOR_POSITIONING_POLYFILL_OPTIONS` object, which would allow using the new `elements` and `excludeInlineStyles` with the default build, and would match the options object pattern we're using here. It seems pretty straightforward so I'm inclined to do that -- any objections to that idea?

marchbox replied:
Ah nice, no objection here. I think I can incorporate that in this PR.

jgerigmeyer commented on src/polyfill.ts:
```diff
@@ -443,13 +443,50 @@ async function position(rules: AnchorPositions, useAnimationFrame = false) {
-      : animationFrame;
+      : options.useAnimationFrame;
+
+  if (options.elements && !Array.isArray(options.elements)) {
```
```suggestion
  if (!Array.isArray(options.elements)) {
```

jgerigmeyer commented on src/fetch.ts:
```diff
@@ -31,45 +31,64 @@ async function fetchLinkedStylesheets(
-  });
+    });
+
+  const elementsForInlines = excludeInlineStyles
+    ? elements?.length
+      ? elements // Only elements in `elements`
+      : [] // No elements
+    : undefined; // All elements
```
Is there a benefit to checking for `elements?.length`?

```suggestion
  const elementsForInlines = excludeInlineStyles ? elements : undefined;
```

marchbox replied:
This is cover this case:

```js
polyfill({ excludeInlineStyles: true });
```

In this case, the author opted out of auto polyfilling inline styles, but didn't pass in an `elements` array. Without this check, `elements` would be `undefined`, and `fetchInlineStyles()` would polyfill all inline styles.

jgerigmeyer replied:
Ah, I see! I might find it slightly easier to read as:

```suggestion
  const elementsForInlines = excludeInlineStyles ? (elements ?? []) : undefined;
```

But it's also fine as-is. 👍

jgerigmeyer commented on tests/unit/fetch.test.ts:
```diff
@@ -97,3 +97,122 @@ describe('fetch inline styles', () => {
+      el2,
+      el3,
+      el4,
+      // should be ignore
+      el5,
+      // @ts-expect-error should be ignored
+      undefined,
+      // @ts-expect-error should be ignore
```
```suggestion
      // should be ignored
      el5,
      // @ts-expect-error should be ignored
      undefined,
      // @ts-expect-error should be ignored
```

jgerigmeyer reviewed (commented):
I love this set of changes -- thanks again for the contribution!

Looks like there are two failing tests. I left mostly minor nit comments, and one question/suggestion for expanding the options API.

jgerigmeyer commented:
> * If you click the imperative button, and then apply the polyfill to the whole page, the polyfill crashes. This is somewhat related to [[BUG] Polyfill fails if a `<link>` fails to load or returns non-CSS content #255](https://github.com/oddbird/css-anchor-positioning/issues/255) as well as [applying the polyfill to dynamic elements #91](https://github.com/oddbird/css-anchor-positioning/issues/91), so I'm not concerned here.

I'm seeing this too. I think it relates to #253, as the error I see seems to be when the polyfill tries to fetch the already-polyfilled style:

```
Security Error: Content at http://localhost:3000/ may not load data from blob:http://localhost:3000/8749e98f-f55a-4153-9249-f6bba301e298.
```

I'm not sure if there's an easy way to read/fetch a blob url, or whether we should explicitly just skip those links?

marchbox commented:
> > * If you click the imperative button, and then apply the polyfill to the whole page, the polyfill crashes. This is somewhat related to [[BUG] Polyfill fails if a `<link>` fails to load or returns non-CSS content #255](https://github.com/oddbird/css-anchor-positioning/issues/255) as well as [applying the polyfill to dynamic elements #91](https://github.com/oddbird/css-anchor-positioning/issues/91), so I'm not concerned here.
> 
> I'm seeing this too. I think it relates to #253, as the error I see seems to be when the polyfill tries to fetch the already-polyfilled style:
> 
> ```
> Security Error: Content at http://localhost:3000/ may not load data from blob:http://localhost:3000/8749e98f-f55a-4153-9249-f6bba301e298.
> ```
> 
> I'm not sure if there's an easy way to read/fetch a blob url, or whether we should explicitly just skip those links?

Not quite a fix but this would at least let other styles being polyfilled:

```ts
try {
  // fetch css and add to array
  const response = await fetch(data.url.toString());
  const css = await response.text();
  return { ...data, css } as StyleData;
} catch {
  return { ...data, css: ''} as StyleData;
}
```

marchbox commented:
@jamesnw  @jgerigmeyer Thanks for the reviews, both. Have addressed all comments. Ready for another round :)

jgerigmeyer reviewed (approved):
> Not quite a fix but this would at least let other styles be polyfilled:
> 
> ```ts
> try {
>   // fetch css and add to array
>   const response = await fetch(data.url.toString());
>   const css = await response.text();
>   return { ...data, css } as StyleData;
> } catch {
>   return { ...data, css: ''} as StyleData;
> }
> ```

I think this might already be fixed by #258? Regardless, I don't think we need to address that in this PR.

Thanks again for this contribution! 🎉

jamesnw reviewed (approved):
This works great! @marchbox I opened PR #259  to update the docs with the new configuration options- I'd appreciate your eyes on that to make sure it makes sense. Thanks!

marchbox commented:
> This works great! @marchbox I opened PR #259 to update the docs with the new configuration options- I'd appreciate your eyes on that to make sure it makes sense. Thanks!

Thank you! Will take a look at the documentation changes today.
